### DISCUSSION TITLE
peer-checked with dynamic values

### DISCUSSION THREAD

**Discussion by ghost** [tailwindlabs/tailwindcss#10445] [Help]
Hi there,

I'm having issues where i'm not able to use a dynamically assigned value with peer-checked. I'm defining a variable for a specific product color like this:

`{% capture bgColor %}bg-products-{{ product.metafields.color.color }}{% endcapture %}` (output = bg-products-green-1)

My tailwind.config.js has this:

`module.exports = {
    mode: 'jit',
    darkMode: false,
    theme: {
        colors: {
            products: {
                'green-1': '#47a35c', //Grüne Mutter
		...`

Inside my element, i'm using it like this:

`peer-checked:{{bgColor}}`

When i'm replacing {{bgColor}} with a default color, like bg-gray-400, it works. 
Using {{bgColor}} throughout my project without any issues for all sorts of applications (like hover etc).

Strangely enough, i've seen it working with the {{ bgColor }} value but after updating my project with npm it's gone again.

Any insights?

**Comment by wongjn** [comment]
From [the documentation](https://tailwindcss.com/docs/content-configuration#dynamic-class-names):
> The most important implication of how Tailwind extracts class names is that it will only find classes that exist as complete unbroken strings in your source files.
>
> If you use string interpolation or concatenate partial class names together, Tailwind will not find them and therefore will not generate the corresponding CSS

**Reply by ghost** [reply]
But i am using "unbroken" values, as {{ bgColor }} equals "bg-products-green-1". Or am i misinterpreting?

**Reply by wongjn** [reply]
Yes, you are misinterpreting. The code needs to be `peer-checked:bg-products-green-1` not `peer-checked:{{bgColor}}`. The former class exists as an unbroken string whereas the latter *is* a broken class string, since the `bg-products-green-1` portion comes from somewhere else.

**Reply by ghost** [reply]
Unfortunately, this doesn't work either. I've now set it like this:

`{% capture peerCheckedBgColor %}peer-checked:bg-products-{{ product.metafields.color.color }}{% endcapture %}`

Using like this:

`class="{{ peerCheckedBgColor }}"`

**Reply by wongjn** [reply]
Yes, it's still a broken string and you are still using string interpolation. It needs to be something like `peer-checked:bg-products-green-1`, not `peer-checked:bg-products-{{ product.metafields.color.color }}`.

**Reply by ghost** [reply]
I understand. The funny thing is that i now suddenly works again, using the same method as before. Pretty difficult to troubleshoot this. Can't be cache either since i've just tested in 3 browsers. Huh..

**Comment by ghost** [comment]
It's not working anymore so back to square one. As far as i see it, there is no way for me to dynamically set the color based on a certain product value. Is there no other way to accomplish what i want?

**Reply by wongjn** [reply]
Here are some possible solutions, (I'm going to use Twig syntax since it seems to be similar to the code syntax you have used):

## Class Map

Have a mapping of keys to class names. This file that this map resides in needs to be covered by the `content` file globs, if you separate it from the template:

```twig
{%
  set color_map = {
    'green-1': '… peer-checked:bg-produces-green-1 …',
    'green-2': '… peer-checked:bg-produces-green-2 …',
    …
  },
%}

{# product.metafields.color.color could be 'green-1', 'green-2', etc. #}
<div class="… {{ color_map[product.metafields.color.color] }} …">
```

The advantage here is that you can use the colors from your Tailwind configuration, but if there are many options, this can get inconvenient to update/write out.

## CSS Variables

If it could be agreeable to store the product color as a valid [CSS `<color>` value](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value) in the datasource, then you could set the color value dynamically using an arbitrary variant (or it could be abstracted to configuration):

```twig
{# product.metafields.color.color could be '#123456', 'red', 'rgb(255, 50, 123)', etc. #}
<div class="… peer-checked:bg-[color:var(--color)] …" style="--color: {{ product.metafields.color.color }};">
```

This has the advantage that you wouldn't need to write out every single combination and less classes might be generated, but you wouldn't be able to directly reference colors from your Tailwind configuration in the datasource values.

## Safelist

You could use the [safelist](https://tailwindcss.com/docs/content-configuration#safelisting-classes) configuration to tell Tailwind to always generate some classes:

```js
module.exports = {
  // …
  safelist: [
    'peer-checked:bg-products-green-1',
    'peer-checked:bg-products-green-2',
  ],
  // …
}
```
Or to cover a wider range of classes without having to maintain a static list:

```js
module.exports = {
  // …
  safelist: [
    {
      pattern: /^bg-products-/,
      variants: ['peer-checked'],
    },
  ],
  // …
}
```

This would mean your existing code will work (`bg-products-{{ product.metafields.color.color }}` or `peer-checked:{{bgColor}}`) but this may generate a lot more classes that you're actually using and could bloat the CSS file size.

**Reply by ghost** [reply]
Thank you wongjn! I started fiddling with the safelist idea and finally did it like this:

```
plugins: [
        require('tailwind-safelist-generator')({
            patterns: [
		'peer-checked:bg-products-{colors.products}',
```

This works out of the box! Thank you!
